flink-core/src/main/java/org/apache/flink/api/connector/source/SourceReaderContext.java

```diff
@@ -63,4 +63,11 @@ public interface SourceReaderContext {
      * @see UserCodeClassLoader
      */
     UserCodeClassLoader getUserCodeClassLoader();
+
+    /**
+     * Get the current parallelism of this Source.
+     *
+     * @return the parallelism of the Source.
+     */
+    int currentParallelism();
```

Comment:
This breaks source compatibility. Either justify an exclusion or add a default implementation.

Reply:
Discussed with Becket [on the ML](https://lists.apache.org/thread/73t5mzsjyhtho9827k0pkqxmocp236xz).
Justified/described in the FLIP [here](https://cwiki.apache.org/confluence/display/FLINK/FLIP-238%3A+Introduce+FLIP-27-based+Data+Generator+Source#:~:text=In%20order%20to,the%20currentParallelism()%20method%3A).
Do you think adding the default method is the right way?

Reply:
I'm not sure.

It seems unlikely that there are implementations of this class, and if there are they should be notified about this additional method in case and reader they use was updated to already rely on it.

While adding a default method is neat in terms of source compatibility, you either:
a) return some static count, which might break behavior (maybe right away, maybe at some point in the future)
b) thrown an exception, which would make it technically source compatible (urgh) but not in practice.

I'm just annoyed that we yet again failed to wrap all this super common metadata (parallelism, attempt number etc etc) into a common structure that can be shared across the reader context and runtime context, and now end up having to copy things over one getter at a time.

Reply:
>I'm just annoyed that we yet again failed to wrap all this super common metadata (parallelism, attempt number etc etc) into a common structure that can be shared across the reader context and runtime context, and now end up having to copy things over one getter at a time.

I see, that is indeed not optimal. Seems like we have three main options:

1. Announce breaking the compatibility and just add it as is. This could be OK for now, but, as you said, more such getters might be needed to be added later, and doing this every time is not great.
2. Add a default method that throws an exception, since, it seems, there is no reasonable default for it.
3. Start a larger effort: create the list of all "info" fields of the `[Stream]RuntimeContext`, pick all that might be useful and expose them as a read-only context object. This will help with supporting the future use cases that are not yet evident, but breaks the encapsulation.

Reply:
I tend to favor the second option.  @zentol what do you think?

Reply:
Let's go with 2).

Reply:
Done.